src/internal/pfsdb/repos.go

```diff
@@ -0,0 +1,363 @@
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return nil, ErrRepoNotFound{ID: id}
+		}
+		return nil, errors.Wrap(err, "scanning repo row")
+	}
+	return getRepoFromRepoRow(row)
+}
+
+// GetRepoByName retrieves an entry from the pfs.repos table by project, repo name, and type.
+func GetRepoByName(ctx context.Context, tx *pachsql.Tx, repoProject, repoName, repoType string) (*pfs.RepoInfo, error) {
+	row, err := getRepoRowByName(ctx, tx, repoProject, repoName, repoType)
+	if err != nil {
+		return nil, err
+	}
+	return getRepoFromRepoRow(row)
+}
+
+func getRepoFromRepoRow(row *repoRow) (*pfs.RepoInfo, error) {
+	proj := &pfs.Project{
+		Name: row.ProjectName,
+	}
+	branches, err := getBranchesFromRepoRow(row)
+	if err != nil {
+		return nil, errors.Wrap(err, "getting branches from repo row")
+	}
+	repoInfo := &pfs.RepoInfo{
+		Repo: &pfs.Repo{
+			Name:    row.Name,
+			Type:    row.RepoType,
+			Project: proj,
+		},
+		Description: row.Description,
+		Branches:    branches,
+		Created:     timestamppb.New(row.CreatedAt),
+	}
+	return repoInfo, nil
+}
+
+func getBranchesFromRepoRow(row *repoRow) ([]*pfs.Branch, error) {
+	if row == nil {
+		return nil, errors.Wrap(fmt.Errorf("repo row is nil"), "")
+	}
+	branches := make([]*pfs.Branch, 0)
+	if row.Branches == noBranches {
+		return branches, nil
+	}
+	// after aggregation, braces, quotes, and leading hex prefixes need to be removed from the encoded branch strings.
+	for _, branchStr := range strings.Split(strings.Trim(row.Branches, "{}"), ",") {
+		branchHex := strings.Trim(strings.Trim(branchStr, "\""), "\\x")
+		decodedString, err := hex.DecodeString(branchHex)
+		if err != nil {
+			return nil, errors.Wrap(err, "branch not hex encoded")
+		}
+		branchInfo := &pfs.BranchInfo{}
+		if err := proto.Unmarshal(decodedString, branchInfo); err != nil {
+			return nil, errors.Wrap(err, "error unmarshalling")
+		}
+		branches = append(branches, branchInfo.Branch)
+	}
+	return branches, nil
+}
+
+// UpsertRepo will attempt to insert a repo. If the repo already exists, it will update its description.
+func UpsertRepo(ctx context.Context, tx *pachsql.Tx, repo *pfs.RepoInfo) error {
+	if repo.Repo.Type == "" {
+		repo.Repo.Type = "unknown"
+	}
+	if repo.Repo.Project == nil {
+		repo.Repo.Project = &pfs.Project{Name: "default"}
+	}
+	_, err := tx.ExecContext(ctx, "INSERT INTO pfs.repos (name, type, project_id, description) "+
+		"VALUES ($1, $2, (SELECT id from core.projects where name=$3), $4) "+
+		"ON CONFLICT (name, type, project_id) DO UPDATE SET description=$5;",
+		repo.Repo.Name, repo.Repo.Type, repo.Repo.Project.Name, repo.Description, repo.Description)
+	if err != nil {
+		return errors.Wrap(err, "upsert repo")
+	}
+	return nil
+}
+
+// UpdateRepo overwrites an existing repo entry by pachsql.ID.
+func UpdateRepo(ctx context.Context, tx *pachsql.Tx, id pachsql.ID, repo *pfs.RepoInfo) error {
+	if repo.Repo.Type == "" {
+		repo.Repo.Type = "unknown"
```

Comment:
This is essentially a change to the data model.  Is there a design doc where I can read about repo types, or any more context for this?

Reply:
That's unintentional on my part. I wanted to added a 'sane' default because leaving this empty was leading to a panic during development. We can omit these defaults if we think that's better.